ansible/devutils

```diff
@@ -18,80 +30,136 @@ def run_cmd(cmd):
+        for hostname, vars in hosts.items():
+            args = {'hostname': vars['ansible_host'],
+                    'username': vars['creds']['username'],
+                    'passwords': vars['creds']['password'],
+                    'cmd': cmd
+                    }
+            g_task_runner.submit_task(hostname, ssh_run_command, **args)
+        for name, result in g_task_runner.task_results():
+            print("task result for {} ===============>\n{}".format(name, str(result['result'][1])))
+
+def ssh_run_command(hostname, username, passwords, cmd):
+    client = SSHClient()
+    client.connect(hostname=hostname, username=username, passwords=passwords)
+    return client.run_command(cmd)
+
+def validate_args(args):
+    if args.action == 'run' and args.cmd == '':
+        print("command is missing for run action")
+        return False
+    return True
 
 def main():
     parser = argparse.ArgumentParser(description='Device utilities')
     parser.add_argument('-6', '--ipv6', help='Include IPv6', action='store_true',
                         required=False, default=False)
     parser.add_argument('-a', '--action',
-                        help='Action towards host(s): list, ping, ssh, console, default list',
+                        help='Action towards host(s): list, ping, run, ssh, console, default list',
                         type=str, required=False, default='list',
-                        choices=['list', 'ping', 'ssh', 'console'])
-    parser.add_argument('-c', '--category', help='Categories: all, sonic, ptf, pdu, default all',
+                        choices=['list', 'ping', 'ssh', 'console', 'run'])
+    parser.add_argument('--cmd', help="Command to run on all hosts",
+                        type=str, required=False)
+    parser.add_argument('-g', '--group', help='Groups: all, sonic, ptf, pdu, default all',
                         type=str, required=False, default='all')
     parser.add_argument('-i', '--inventory', help='Categories: lab, etc, default lab',
                         type=str, required=False, default='lab')
     parser.add_argument('-l', '--limit', help='Host: limit to a single dut host name, default all',
-                        type=str, required=False, default='all')
+                        type=str, required=False)
     parser.add_argument('-u', '--user', help='User: user account to login to host with, default admin',
                         type=str, required=False, default='admin')
+    parser.add_argument('-c', '--concurrency', help='Concurrency: the max concurrency for tasks that can run simultaneously, default 1',
+                        type=int, required=False, default=1)
 
     args = parser.parse_args()
-
-    hosts = get_host_list(args.inventory, args.category)
+    if not validate_args(args):
+        return
+    global g_task_runner, g_inv_mgr
+    g_task_runner = TaskRunner(max_worker=args.concurrency)
+    g_inv_mgr = HostManager(args.inventory)
```

Comment:
Lack handling code when no ansible?

Reply:
Yes. The code without ```Ansible``` is not implemented yet. I added a todo here.